File: cpp/2021/day12/day12.cpp
#include "../../common.h"
#include <map>
#include <string>
#include <deque>
#include <algorithm>

bool is_small(const string &node);

struct path { 
    vector<const string*> nodes;
    const string* duped;
    bool has_dupe;

    path() {
        has_dupe = false;
    }
    path(vector<const string*> &n) {
        nodes = n;
        has_dupe = false;
    }
    path(const path* p) {
        nodes = p->nodes;
        has_dupe = p->has_dupe;
        duped = p->duped;
    }
    void push_back(const string* node) {
        nodes.push_back(node);
        if (!has_dupe && is_small(*node) && count(nodes.begin(), nodes.end(), node) == 2) {
            has_dupe = true;
            duped = node;
        }
    }
    bool node_is_allowed(const string* node) {
        if (!is_small(*node)) return true;
        return (!has_dupe || (has_dupe && node != duped && count(nodes.begin(), nodes.end(), node) != 1));
    }
};

struct pathfinder {
    map<string, vector<const string*>>* caves;
    deque<path*> q;
    unsigned int num_paths;

    pathfinder(map<string, vector<const string*>>* caves, const string* start_ptr) {
        this->caves = caves;
        path* p = new path();
        p->push_back(start_ptr);
        deque<path*> q;
        q.push_back(p);
        this->q = q;
        num_paths = 0;
    }

    void find_paths() {
        while (!q.empty()) {
            if (q.size() != 0 && **(q.front()->nodes.end()-1) == "end") {
                num_paths++;
                //delete q.front();
                q.pop_front();
                continue;
            }
            if (q.empty()) return;
            for (const string* node : (*caves)[**(q.front()->nodes.end()-1)]) {
                if (!q.front()->node_is_allowed(node)) continue; 
                path* temp = new path(q.front());
                temp->push_back(node);
                q.emplace_back(temp);
            }
            q.pop_front();
        }
    }
};

int main() {
    map<string, vector<const string*>> caves;
    map<string, const string*> pointers;
    
    ifstream infile;
    infile.open("day12.in");

    string line;

    while (getline(infile, line)) {
        int dash_pos = line.find('-');
        const string* first = new string(line.substr(0, dash_pos));
        const string* second = new string(line.substr(dash_pos+1));

        if (pointers.find(*first) == pointers.end()) {
            pointers[*first] = first;
        }
        if (pointers.find(*second) == pointers.end()) {
            pointers[*second] = second;
        }
        if (caves.find(*first) == caves.end()) {
            vector<const string*> temp;
            caves[*first] = temp;
        }
        if (caves.find(*second) == caves.end()) {
            vector<const string*> temp;
            caves[*second] = temp;
        }

        if (*second != "start") caves[*first].push_back(pointers[*second]);
        if (*first != "start") caves[*second].push_back(pointers[*first]);
    }

    caves.erase("end");

    infile.close();

    // for (const auto &[key, value] : pointers) {
    //     cout << key << '=' << *value << endl;
    // }

    // for (const auto &[key, value] : caves) {
    //     cout << key << " = { ";
    //     for (auto v : value) {
    //         cout << *v << " ";
    //     }
    //     cout << "}" << endl;
    // }

    pathfinder pathf(&caves, pointers["start"]);

    pathf.find_paths();

    LOG(pathf.num_paths);
}

bool is_small(const string &node) {
    return (node[0] >= 'a' && node[0] <= 'z');
}

File: csharp/2021/day12/day12.cs
// I'm making this with little experience and no internet, so this won't look pretty.
using System;
using System.IO;
using System.Collections.Generic;
using System.Linq;

namespace Advent2021
{
    class Day12
    {
        static void Main(string[] args)
        {   
            var caves = new List<Cave>();
            foreach (string line in File.ReadAllLines("./day12.in"))
            {
                string[] connection = line.Split("-");
                Cave first = new Cave(connection[0]);
                Cave second = new Cave(connection[1]);

                if (!caves.Contains(first))
                {
                    caves.Add(first);
                }
                if (!caves.Contains(second))
                {
                    caves.Add(second);
                }

                if (second.Name != "start") caves[caves.IndexOf(first)].AddNeighbor(second);
                if (first.Name != "start") caves[caves.IndexOf(second)].AddNeighbor(first);
            }

            // foreach (var cave in caves)
            // {
            //     Console.Write($"{cave};{cave.IsSmall} = ");
            //     foreach (var n in cave.Neighbors)
            //     {
            //         Console.Write($"{n};{n.IsSmall} ");
            //     }
            //     Console.WriteLine();
            // }

            var pathfinder = new Pathfinder(caves);

            pathfinder.FindPaths(); 

            Console.WriteLine(pathfinder.NumPaths);
        }
    }
    class Pathfinder
    {
        private List<Cave> caves;
        private Queue<Route> queue;
        private int numPaths;

        public Pathfinder(List<Cave> caves)
        {
            this.caves = caves;
            queue = new Queue<Route>();
            queue.Enqueue(new Route(new Cave("start")));
        }

        public void FindPaths()
        {   
            //this whole time... the issue could have been resolve by doing a while loop instead of recursion...
            while (queue.Count() != 0) 
            {
                if (queue.Peek().Top().Name == "end")
                {
                    numPaths++;
                    queue.Dequeue();
                    continue;
                }
                foreach (Cave node in caves[caves.IndexOf(queue.Peek().Top())].Neighbors)
                {
                    if (!queue.Peek().IsNodeAllowed(node)) continue;
                    Route temp = new Route(queue.Peek());
                    temp.AddNode(node);
                    queue.Enqueue(temp);
                }
                queue.Dequeue();
            }
        }

        public int NumPaths
        {
            get { return numPaths; }
        }

    }
    class Route : IEquatable<Route>
    {
        private List<Cave> nodes;
        private string dupe;

        public Route()
        {
            nodes = new List<Cave>();
            dupe = "";
        }
        public Route(Route r) 
        {
            dupe = r.dupe;
            nodes = new List<Cave>(r.nodes);
        }
        public Route(Cave start) : this()
        {
            AddNode(start);
        }

        public Cave Top()
        {
            return nodes[nodes.Count()-1];
        }

        public void AddNode(Cave node)
        {
            nodes.Add(node);
            if (!HasDupe && node.IsSmall && nodes.Where(cave => cave.Equals(node)).Count() == 2)
            {
                dupe = node.Name;
            }
        }

        public bool IsNodeAllowed(Cave node)
        {
            if (!node.IsSmall) return true;
            return (!HasDupe || (HasDupe && node.Name != dupe && nodes.Where(cave => cave.Equals(node)).Count() != 1));
        }

        public virtual bool Equals(Route r)
        {
            return r.nodes.Equals(nodes);
        }

        public bool HasDupe
        {
            get { return dupe != ""; }
        }
        public List<Cave> Nodes
        {
            get { return nodes; }
        }
    }

    class Cave
    {
        private string name;
        private List<Cave> neighbors;

        public Cave(string name) 
        {
            this.name = "" + name;
            neighbors = new List<Cave>();
        }

        public void AddNeighbor(Cave n)
        {
            neighbors.Add(n);
        }

        public List<Cave> Neighbors
        {
            get { return neighbors; }
        }

        public bool IsSmall
        {
            get { return (name == name.ToLower()); }
        }

        public string Name
        {
            get { return name; }
        }

        public override string ToString()
        {
            return name;
        }

        public override bool Equals(object obj)
        {
            if (obj == null || GetType() != obj.GetType())
            {
                return false;
            }
            
            return name == ((Cave) obj).name;
        }
        
        public override int GetHashCode()
        {
            return name.GetHashCode();
        }
    }
}